marketplace-service/src/main/java/com/axonivy/market/repository/impl/ProductListedRepositoryImpl.java

```diff
@@ -0,0 +1,145 @@
+
+  private final MongoTemplate mongoTemplate;
+
+  public ProductListedRepositoryImpl(MongoTemplate mongoTemplate) {
+    this.mongoTemplate = mongoTemplate;
+  }
+
+  @Override
+  public Page<Product> findAllListed(Pageable pageable) {
+    var criteria = new Criteria();
+    criteria.andOperator(createListedFilter().ne(false));
+    return findByCriteria(pageable, criteria);
+  }
+
+  @Override
+  public Page<Product> searchListedByCriteria(ProductSearchCriteria searchCriteria, Pageable pageable) {
+    return findByCriteria(pageable, buildCriteriaSearch(searchCriteria, true, false));
+  }
+
+  @Override
+  public Product findListedByCriteria(ProductSearchCriteria criteria) {
+    Criteria searchCriteria = buildCriteriaSearch(criteria, true, false);
+    List<Product> entities = mongoTemplate.find(new Query(searchCriteria), Product.class);
+    return CollectionUtils.isEmpty(entities) ? null : entities.get(0);
+  }
+
+  @Override
+  public Product findByCriteria(ProductSearchCriteria criteria) {
+    Criteria searchCriteria = buildCriteriaSearch(criteria, false, false);
+    List<Product> entities = mongoTemplate.find(new Query(searchCriteria), Product.class);
+    return CollectionUtils.isEmpty(entities) ? null : entities.get(0);
+  }
+
+  private Page<Product> findByCriteria(Pageable pageable, Criteria criteria) {
+    var query = new Query(criteria);
+    query.with(pageable);
+    List<Product> entities = mongoTemplate.find(query, Product.class);
+    long count = mongoTemplate.count(new Query(criteria), Product.class);
+    return new PageImpl<>(entities, pageable, count);
+  }
+
+  private Criteria createListedFilter() {
+    return Criteria.where(LISTED.getFieldName());
+  }
+
+  private Criteria buildCriteriaSearch(ProductSearchCriteria searchCriteria, boolean isListed,
+      boolean isWrappedQueryByOr) {
+    var criteria = new Criteria();
+    List<Criteria> andFilters = new ArrayList<>();
+    List<Criteria> orFilters = new ArrayList<>();
+
+    // Query by Listed
+    if (isListed) {
+      Criteria listedCriteria = createListedFilter();
+      listedCriteria.ne(false);
+      andFilters.add(listedCriteria);
+    }
+
+    // Query by Type
+    if (searchCriteria.getType() != null && TypeOption.ALL != searchCriteria.getType()) {
+      Criteria typeCriteria = Criteria.where(TYPE.getFieldName());
+      andFilters.add(typeCriteria.is(searchCriteria.getType().getCode()));
+    }
+
+    // Query by Keyword regex
+    if (StringUtils.isNoneBlank(searchCriteria.getKeyword())) {
+      createQueryByKeywordRegex(searchCriteria, orFilters);
+    }
+
+    // Add OR operator is root condition
+    if (isWrappedQueryByOr && !CollectionUtils.isEmpty(orFilters)) {
+      criteria.orOperator(orFilters);
+      if (!CollectionUtils.isEmpty(andFilters)) {
+        criteria.andOperator(andFilters);
+      }
+    } else {
+      // Add AND operator is root condition
+      if (!CollectionUtils.isEmpty(andFilters)) {
```

Comment:
check logic here, seems like it not works

Reply:
can you give me a test case that is not work?
this is a main block I tested, and I can confirm it worked as design